common/dpcpp/slab_hash.hpp

```diff
@@ -0,0 +1,182 @@
+        _val = val;
+
+        if (_ind == 0) {
+            _iter = (_lists + _hasher(key))->root;
+        }
+        sycl::group_barrier(_gr);
+
+        while (_iter != nullptr) {
+            if (insert_in_node()) {
+                break;
+            } else if (_ind == 0) {
+                _iter = _iter->next;
+            }
+
+            sycl::group_barrier(_gr);
+        }
+    }
+
+    pair<T, bool> find(K key) {
+        _key = key;
+        _ans = {T(), false};
+        
+        if (_ind == 0) {
+            _iter = (_lists + _hasher(key))->root;
+        }
+        sycl::group_barrier(_gr);
+
+        while (_iter != nullptr) {
+            if (find_in_node()) {
+                break;
+            } else if (_ind == 0) {
+                _iter = _iter->next;
+            }
+
+            sycl::group_barrier(_gr);
+        }
+
+        return _ans;
+    }
+
+private:
+    bool insert_in_node() {
+        bool total_found = false;
+        bool find = false;
+
+        for(int i = _ind; i <= _ind + SUBGROUP_SIZE * (CONST - 1); i += SUBGROUP_SIZE) {
+            find = ((_iter->data[i].first) == _empty);
+            sycl::group_barrier(_gr);
+            total_found = sycl::any_of_group(_gr, find);
+
+            if (total_found) {
+                if (insert_in_subgroup(find, i)) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    bool insert_in_subgroup(bool find, int i) {
+        for (int j = 0; j < SUBGROUP_SIZE; j++) {
+            if (cl::sycl::group_broadcast(_gr, find, j)) {
+                K tmp_empty = _empty;
+                bool done = _ind == j ? sycl::ONEAPI::atomic_ref<K, sycl::ONEAPI::memory_order::acq_rel,
+                                            sycl::ONEAPI::memory_scope::system,
+                                            sycl::access::address_space::global_space>(
+                                                _iter->data[i].first
+                                            ).compare_exchange_strong(tmp_empty,
+                                                                        _key) : false;
+                if (done) {
+                    _iter->data[i].second = _val;
+                }
+                if (cl::sycl::group_broadcast(_gr, done, j)) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    bool find_in_node() {
+        bool find = false;
+        bool total_found = false;
+
+        for(int i = _ind; i <= _ind + SUBGROUP_SIZE * (CONST - 1); i += SUBGROUP_SIZE) {
+            find = ((_iter->data[i].first) == _key);
+            sycl::group_barrier(_gr);
+            total_found = sycl::any_of_group(_gr, find);
+            
+    
+            if (total_found) {
+                find_in_subgroup(find, i);
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    void find_in_subgroup(bool find, int i) {
```

Comment:
Any reason not to return function result?

Reply:
find_in_subgroup is called only if workgroup has found something. In this case, find_in_subgroup will always find the result, so I thought that it has nothing to return.